### PR TITLE
Implement osu!catch star rating

### PR DESCRIPTION
**Pull request by frankhjwx**
The star rating algorithm is transplanted from https://gist.github.com/peppy/beac8bf3d0da2191ef14d420ff1119e4 and allows lazer to support visualizing catch star rating now.

### CODE REVIEW

File: osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs [hunk 1]
```diff
@@ -37,5 +37,20 @@ protected virtual void PreprocessHitObjects()
         }
 
         public abstract double Calculate(Dictionary<string, double> categoryDifficulty = null);
+
+        protected double StrainCalculator(List<double> highestStrains, double decayWeight)
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs]: `CalculateStrain`

Also should have an xmldoc with the first comment: "Finds the weighted sum of a sequence of strain intervals."

**Review by smoogipoo — DISMISSED** [review]

**Review by DrabWeb — CHANGES_REQUESTED** [review]
Please don't just copy-paste the stable reference, it's a reference for a reason.

File: osu.Game.Rulesets.Catch/Difficulty/CatchDifficultyHitObject.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using System;
+using osu.Game.Rulesets.Catch.Objects;
+using osu.Game.Rulesets.Catch.UI;
+using OpenTK;
+
+namespace osu.Game.Rulesets.Catch.Difficulty
+{
+    internal class CatchDifficultyHitObject
```
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game.Rulesets.Catch/Difficulty/CatchDifficultyHitObject.cs]: Given that this is a direct re-implementation of osu!stable's calculation, I wouldn't be too harsh about this. The entire calculator should/will be reworked in the future anyway.

**Comment by smoogipoo** [comment]
@peppy Requesting a review on this since I've made catcher changes.

It seems like the size is not quite exact to osu!stable (off by 20-40 pixels). At least one contributor is the missing 0.1f catch margins, but I'm not sure if they should be implemented anyway/what their purpose was.

It's a start at least.

**Review by peppy — APPROVED** [review]

**Pull request merged**